README.md
<p align="center">
  <a href="https://travis-ci.com/konkor/obmin"><img src="https://travis-ci.com/konkor/obmin.svg?branch=master" alt="TravisCI"></a>
  <a href="https://github.com/konkor/obmin"><img src="https://img.shields.io/github/license/konkor/obmin.svg" alt="GPLv3 License"></a>
  <a href="https://github.com/konkor/obmin"><img src="https://img.shields.io/github/stars/konkor/obmin.svg?style=social&label=Star&style=flat-square" alt="Stars"></a><br>
</p>

<p align="center">
<a href="https://obmin.github.io"><img src="/data/icons/obmin-run.png" alt="LOGO"></a>
</p>

### [OBMIN](https://obmin.github.io/) **One-Click** File Sharing Solution for your home network and worldwide.
-----

<h2 align="center">Supporting OBMIN Server</h2>

OBMIN Server is an GPLv3-licensed open source project focused on desktop users. It's an independent project with its ongoing development made possible entirely thanks to the support by these awesome [backers](https://github.com/konkor/obmin/blob/master/BACKERS.md). You will also support my other interesting community projects. If you'd like to help the project and want to join to it, please consider:

- [Become a backer or sponsor on Patreon](https://www.patreon.com/konkor).
- [One-time donation via PayPal or Patron.](#donations)

Please, consider to support the project and make it better!

<h3 align="center">Special Sponsors</h3>
<!--special start-->

<p align="center">
  <big>
    Top supporters:<br><br><b>
    <a href="https://obmin.github.io/supporters">Jonathan Alden</a><br>
    <a href="https://obmin.github.io/supporters">Manuel Transfeld</a><br>
    <br></b>
    <a href="https://obmin.github.io/supporters">Joël Dinel</a><br>
    <a href="https://obmin.github.io/supporters">Michel Bonifert</a><br>
    <a href="https://obmin.github.io/supporters">Patrick Strobel</a><br>
    <br>
    Patreons:<br><br>
    <a href="https://github.com/krzemienski">Nick Krzemienski (USA)</a><br>
    <a href="https://vk.com/anaumynaugames">Ivan Chayka (Russia)</a><br>
    <a href="https://obmin.github.io/supporters">Steffen W. (Germany)</a><br>
    <a href="https://obmin.github.io/supporters">Yoann Deferi (France)</a><br>
    <br>
    <a href="https://obmin.github.io/supporters">More wonderful people...</a><br>
    <br>
    <i>The one top will be placed in UI like 'Sponsored by ...' you could be here with your name or logo and reference (contact me or put links in donation's messages)</i>
  </big>
</p>

<!--special end-->

![screencast](https://user-images.githubusercontent.com/1944781/27997375-a73383c2-64ff-11e7-8a86-b9fddca45f42.png)
## Introduction
**Obmin** has convenient user-friendly systray applets for many Desktop Environments like Gnome Shell, Unity, KDE, Mate, LXDE, XFCE, Cinnamon, Pantheon, Budgie DE through the **obmin-indicator** and [Gnome Obmin Extension](https://extensions.gnome.org/extension/1254/obmin/).

## Main Features
* Easy installation.
* Easy setup just choose file(s) locations and click Obmin on.
* Doesn't require ROOT privileges.
* Doesn't require any special client side installation.
* HTTP(S) transfer protocol available everywhere Linux, OSX, Windows, Android, iOS so.
* Secure HTTPS connections.

## Enchanted Features
* Multiple file sources supporting.
* Content filters: symbolic links, backups, hidden files.
* Port configuration.
* TLS Certificate generating.
* Logging of detailed server connections information.
* Real-time monitor of incoming connections and traffic.
* Content theming.
* Various Linux Desktop Environments and distributions.
* Obmin extensions like compressor, slideshow, playlist generator.
* Real-time video streamer with build-in remuxing and encoding.

## Planned Features
* Project website (GitHub Pages).
* Improve logging and monitors.
* More useful extensions.
* Online service to enhance obmin connectivity (need donations).

## Online Resources
* [Project Site](https://obmin.github.io)
* [Wiki](https://github.com/konkor/obmin/wiki)
* [Docker](https://hub.docker.com/r/obmin/obmin/)
* [Gnome](https://extensions.gnome.org/extension/1254/obmin/)

## Contributions
* Report [a bug](https://github.com/konkor/obmin/issues).
* Test it on your favorite Linux distribution.
* Contribute with an idea, graphical content, translation or code [an issue](https://github.com/konkor/obmin/issues).
* Make donation to the project.

## Donations
  I like completely open-source projects. It's why I picked GPLv3 license for my open projects. I think only such license could protect Desktop Users from Business Users. Maybe I'm a dreamer and want to believe in the pure projects but the reality is most projects and FOS organizations are sponsored by big business and founded by them.
  But real life is a hard thing and very complicated by many circumstances. I'm not young and all we have it's our life (time) and where we'll be tomorrow. Life is hard. I'd like it to work on my projects productively which want a lot of time and affords. Now I want get your support to have ability to support and develop projects.

- [Become a backer or sponsor on Patreon](https://www.patreon.com/konkor).
- One-time donation via [PayPal EURO](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=WVAS5RXRMYVC4), [PayPal USD](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=HGAFMMMQ9MQJ2) or Patron where you can choose a custom pledge.

* Contact to the author [here](https://konkor.github.io/index.html#contact).

_Behind the development for the Linux Desktop are ordinary people who spend a lot of time and their own resources to make the Linux Desktop better. Thank you for your contributions!_


## Installation

[See more about compilation and installation...](/INSTALL.md)

### Sources and binary packages
* [GitHub master branch](https://github.com/konkor/obmin/archive/master.zip)
* [GitHub development branch](https://github.com/konkor/obmin/archive/devel.zip)
* [Gnome extensions repository](https://extensions.gnome.org/extension/1254/obmin/)
* [Latest deb package for Debian/Ubuntu flavors](https://github.com/konkor/obmin/raw/devel/releases/obmin_latest_all.deb)

## License

[GPLv3](https://www.gnu.org/licenses/gpl.html)


debian/changelog
obmin (26-1) UNRELEASED; urgency=low

  * Update fixes for GNOME Shell 43.x series
  * Update ES2015 for server and extension

 -- konkor <kapa76@gmail.com>  Mon, 02 Oct 2023 12:00:00 +0100

obmin (25-1) UNRELEASED; urgency=low

  * Update fixes removing notify events on disable

 -- konkor <kapa76@gmail.com>  Sun, 20 Feb 2022 12:00:00 +0100

obmin (24-1) UNRELEASED; urgency=low

  * Update fixes versioning of JS modules

 -- konkor <kapa76@gmail.com>  Sat, 19 Feb 2022 16:43:00 +0100

obmin (23-1) UNRELEASED; urgency=low

  * Update fixes for GNOME Shell 41x series
  * Fix for GJS Soup.Buffer.new constructor

 -- konkor <kapa76@gmail.com>  Tue, 31 Jan 2022 09:43:00 +0100

obmin (22-1) UNRELEASED; urgency=low

  * Update fixes for GNOME Shell 40x series

 -- konkor <kapa76@gmail.com>  Tue, 25 May 2021 18:58:00 +0100

obmin (21-1) UNRELEASED; urgency=low

  * Add Turkish translations

 -- konkor <kapa76@gmail.com>  Sat, 14 Nov 2020 00:41:00 +0100

obmin (20-1) UNRELEASED; urgency=low

  * Fix for the Digest HTTP Authentication for newer libsoup versions

 -- konkor <kapa76@gmail.com>  Sat, 15 Feb 2020 21:55:00 +0100

obmin (19-1) UNRELEASED; urgency=low

  * Fix for the offending callbacks into JSAPI during the sweeping phase of GC
  * Update minor fixes and improvements

 -- konkor <kapa76@gmail.com>  Fri, 11 Oct 2019 11:57:00 +0100

obmin (18-1) UNRELEASED; urgency=low

  * Next version

 -- konkor <kapa76@gmail.com>  Sat, 06 Apr 2019 13:12:00 +0100

obmin (17-1) UNRELEASED; urgency=low

  * Add support for GNOME Shel 3.32
  * Update styles for control center and minor fixes

 -- konkor <kapa76@gmail.com>  Thu, 04 Apr 2019 12:45:00 +0100

obmin (16-1) UNRELEASED; urgency=low

  * Add HTTP Digest Authentication
  * Add dbus server for messaging

 -- konkor <kapa76@gmail.com>  Fri, 22 Jun 2018 21:14:00 +0100

obmin (15-1) UNRELEASED; urgency=low

  * Add secure https protocol supporting

 -- konkor <kapa76@gmail.com>  Tue, 06 Mar 2018 19:07:00 +0100

obmin (14.0-1) UNRELEASED; urgency=low

  * Update fix for the loading Remembered settings on intel_pstate.
  * Enable Gnome 3.28 supporting

 -- konkor <kapa76@gmail.com>  Wed, 28 Feb 2018 20:47:00 +0100

obmin (13.0-1) UNRELEASED; urgency=medium

  * Move binary/scripts to modules to fix possible issues standalone installed version and updates via gnome extensions.
  * Update minor plugin API
  * Update fix theme styling
  * Add zip archives support to the Compressor extension
  * Add download button to the Slideshow extension
  * Add new Transcoder extension to realtime video content to WEB Browser supported formats (beta - not enabled by default)
  * Add new Camera extension to stream V4L2 Device (beta - not enabled by default)
  * Add OBMIN extension manager for extensions
  * Updated preferences window layout
  * Removed donation buttons from the applets and web ui

 -- konkor <kapa76@gmail.com>  Sun, 03 Dec 2017 17:47:00 +0100

obmin (12.0-1) UNRELEASED; urgency=medium

  * Add slideshow extension
  * Add new widgets class panel
  * Add additional file's filter conditions (internal)
  * Add Chromium supporting for obmin extensions
  * Fix the content padding

 -- konkor <kapa76@gmail.com>  Thu, 26 Oct 2017 00:13:23 +0100

obmin (11.0-1) UNRELEASED; urgency=medium

  * Moduled server architecture (extension API)
  * Filelist extension has recursion
  * New extension for making archives from content folders
  * Various minor fixes and improvements

 -- konkor <kapa76@gmail.com>  Fri, 20 Oct 2017 14:45:00 +0300

obmin (10.0-1) UNRELEASED; urgency=medium

  * Add local logger for the server journal
  * Update theming
  * Minor fixes and updates

 -- konkor <kapa76@gmail.com>  Sat, 14 Oct 2017 21:57:00 +0300

obmin (9.0-1) UNRELEASED; urgency=medium

  * Add obmin control center and monitor
  * Update brazilian portuguese locale, pt_BR
  * Update assets and documentation

 -- konkor <kapa76@gmail.com>  Thu, 12 Sep 2017 18:30:00 +0300

obmin (8.0-1) UNRELEASED; urgency=medium

  * Add activity monitor
  * Add brazilian portuguese locale, pt_BR
  * Update docs, icons, improved playlists
  * Update German locale, de_DE
  * Move the gnome extension's sources to the preferences window
  * Improved stability and performance
  * Minor Fixes

 -- konkor <kapa76@gmail.com>  Sat, 16 Sep 2017 11:45:00 +0300

obmin (7.0-1) UNRELEASED; urgency=medium

  * Add internal counters for requests, responds and uploaded data
  * Update URI playlist escaping

 -- konkor <kapa76@gmail.com>  Mon, 4 Sep 2017 16:48:00 +0300

obmin (6.0-1) UNRELEASED; urgency=medium

  * Add file and playlist
  * Update ES6 standards

 -- konkor <kapa76@gmail.com>  Mon, 4 Sep 2017 16:48:00 +0300

obmin (5.0-1) UNRELEASED; urgency=medium

  * Add configuration via json config file
  * Fix mount points filter
  * Update docs

 -- konkor <kapa76@gmail.com>  Mon, 4 Sep 2017 16:48:00 +0300

obmin (4.0-1) UNRELEASED; urgency=medium

  * Add supporting for non-Gnome
  * Update UI, documentations
  * Fix minnor issues

 -- konkor <kapa76@gmail.com>  Mon, 4 Sep 2017 16:48:00 +0300

obmin (3.0-1) UNRELEASED; urgency=medium

  * Add obmin-indicator launcher for non-Gnome systems with AppIndicator supporting
  * Add full translation supporting
  * Add german translation

 -- konkor <kapa76@gmail.com>  Wed, 29 Aug 2017 22:29:00 +0300

obmin (2.0-1) UNRELEASED; urgency=medium

  * Add bins obmin-server and obmin-preferences

 -- konkor <kapa76@gmail.com>  Tue, 29 Aug 2017 11:45:25 +0300

obmin (1.0-1) UNRELEASED; urgency=medium

  * Initial release

 -- konkor <kapa76@gmail.com>  Mon, 28 Aug 2017 13:45:25 +0300


common/indicator.js
/*
 * This is a part of OBMIN Server
 * Copyright (C) 2017-2019 konkor <konkor.github.io>
 *
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

imports.gi.versions.Gtk = '3.0';

const Gtk = imports.gi.Gtk;
const Gdk = imports.gi.Gdk;
const Gio = imports.gi.Gio;
const GLib = imports.gi.GLib;
const AI = imports.gi.AppIndicator3;
const Lang = imports.lang;

const APPDIR = get_appdir ();
imports.searchPath.unshift(APPDIR);

const Gettext = imports.gettext.domain('gnome-shell-extensions-obmin');
const _ = Gettext.gettext;
const Convenience = imports.convenience;

//const Clipboard = Gtk.Clipboard.get_default(Gdk.Display.get_default());

const HTTPS_KEY = 'https';
const DEBUG_KEY = 'debug';
const STARTUP_KEY = 'startup-settings';
const STATS_MONITOR_KEY = 'stats-monitor';
const SUPPORT_KEY = 'support';
const PORT_KEY = 'port';
const STATUS_KEY = 'status';

let https = false;
let startup = false;
let support = 0;
let port = 8088;
let DEBUG = 1;
let status = 30;
let stats_monitor = true;
let stats = {access:0, ready:0, upload:0};

let status_event = 0;
let update_event = 0;
let settings = null;
let server = 0;
let desktop = '';

var ObminIndicator = new Lang.Class ({
    Name: 'ObminIndicator',

    _init: function () {
        https = settings.get_boolean (HTTPS_KEY);
        DEBUG = settings.get_int (DEBUG_KEY);
        startup = settings.get_boolean (STARTUP_KEY);
        support = settings.get_int (SUPPORT_KEY);
        port = settings.get_int (PORT_KEY);
        status = settings.get_int (STATUS_KEY);
        stats_monitor = settings.get_boolean (STATS_MONITOR_KEY);
        desktop = GLib.getenv ("XDG_CURRENT_DESKTOP").toUpperCase ();

        this.application = new Gtk.Application ({
            application_id: "org.konkor.obmin.indicator",
            flags: Gio.ApplicationFlags.HANDLES_OPEN});
        GLib.set_application_name ("OBMIN Indicator");
        GLib.set_prgname ("OBMIN Indicator");
        this.application.connect ('activate', this._onActivate.bind (this));
        this.application.connect ('startup', this._onStartup.bind (this));
        debug (desktop);
    },

    _onActivate: function (){
        this._window.hide ();
    },

    _onStartup: function () {
        this._window = new Gtk.Window ();
        this._window.title = "OBMIN Indicator";
        this._window.set_icon_name ('obmin');
        if (!this._window.icon) try {
            this._window.icon = Gtk.Image.new_from_file (APPDIR + "/data/icons/obmin.svg").pixbuf;
        } catch (e) {
            error (e.message);
        }
        this.application.add_window (this._window);
        server = this.server_enabled;
        this.build_menu ();
        if (server) {
            this.update_stats ();
        } else if (startup) this._enable (true);
        if (status > 0) status_event = GLib.timeout_add_seconds (0, status, () => {
            this.check_status ();
            return true;
        });
        this.dbus = Gio.bus_get_sync (Gio.BusType.SESSION, null);
        if (stats_monitor && this.dbus)
            this.dbus.call ('org.freedesktop.DBus', '/', "org.freedesktop.DBus", "AddMatch",
                GLib.Variant.new ('(s)', ["type=\'signal\'"]), null, Gio.DBusCallFlags.NONE, -1, null, () => {
                    this._signalCC = this.dbus.signal_subscribe (null, "org.konkor.obmin.server", "CounterChanged",
                    '/org/konkor/obmin/server', null, Gio.DBusSignalFlags.NO_MATCH_RULE, this.on_counter_changed.bind (this));
            });
    },

    on_counter_changed: function (conn, sender, object, iface, signal, param, user_data) {
        stats = JSON.parse (param.get_child_value(0).get_string()[0]);
        if (update_event) GLib.Source.remove (update_event);
        update_event = GLib.timeout_add (0, 50, this.update_stats.bind (this));
    },

    check_status: function () {
        let run = false;
        let res = GLib.spawn_command_line_sync ("ps -A");
        let o, n;
        if (res[0]) o = res[1].toString().split("\n");
        res = null;
        for (let i = 0; i < o.length; i++) {
            if (o[i].indexOf ("obmin-server") > -1) {
                n = parseInt (o[i].trim().split(" ")[0]);
                if (Number.isInteger(n) && n > 0) run = n;
                break;
            }
        }
        debug ("status: " + run?"":"Stopped");
        if (run != server) {
            server = run;
            this.lock = true;
            this.server_switch.set_active (server);
            this.lock = false;
            this.update_icon ();
        }
    },

    update_icon: function () {
        if (!server) this.indicator.set_icon ("obmin-off");
        else if ((stats.access - stats.ready) > 0) {
            if (this.indicator.get_icon() != "obmin-run") this.indicator.set_icon ("obmin-run");
        } else if (this.indicator.get_icon() != "obmin-on") this.indicator.set_icon ("obmin-on");
    },

    update_stats: function () {
        if (update_event) {
            GLib.Source.remove (update_event);
            update_event = 0;
        }
        if (stats.access && (stats.access >= 0)) {
            this.connections.update ((stats.access - stats.ready).toString());
            this.requests.update (stats.access.toString());
            this.uploads.update (GLib.format_size (stats.upload));
        }
        if (server) this.update_icon ();
        else this.indicator.set_icon ("obmin-off");
        return false;
    },

    build_menu: function () {
        var item;
        this.appmenu = new Gtk.Menu ();
        this.indicator = AI.Indicator.new ("Obmin", "obmin-indicator", AI.IndicatorCategory.APPLICATION_STATUS);
        this.indicator.set_icon_theme_path (APPDIR + "/data/icons");
        this.server_switch = Gtk.CheckMenuItem.new_with_label (" ");
        this.server_switch.get_child().set_markup ("<b> Obmin "+_("Server")+"</b>");
        this.server_switch.tooltip_text = _("Activate Obmin Server");
        this.server_switch.active = server;
        this.server_switch.connect ('toggled', () => {
            if (!this.lock) this._enable (this.server_switch.active);
        });
        this.appmenu.append (this.server_switch);
        this.appmenu.append (new Gtk.SeparatorMenuItem ());

        this.info_local = new LocalItem ();
        this.appmenu.append (this.info_local);
        this.info_public = new PublicItem ();
        this.appmenu.append (this.info_public);

        this.stats = Gtk.MenuItem.new_with_label (" ");
        this.stats.get_child().set_markup ("<b>"+_("Usage Statistics")+"</b>");
        this.appmenu.senitive = false;

        this.connections = new InfoMenuItem (_("Active"), "0");
        this.connections.tooltip_text = _("Active connections");
        this.requests = new InfoMenuItem (_("Total Requests"), "0");
        this.requests.tooltip_text = _("Total Number Of Requests To Obmin Server");
        this.uploads = new InfoMenuItem (_("Transferred"), "0 bytes");
        this.uploads.tooltip_text = _("Total Amount Of The Transferred Data From Obmin Server");
        if (desktop != "PANTHEON") {
            this.appmenu.append (new Gtk.SeparatorMenuItem ());
            this.appmenu.append (this.stats);
            this.appmenu.append (this.connections);
            this.appmenu.append (this.requests);
            this.appmenu.append (this.uploads);
            this.connections.connect ('activate', () => {
                GLib.spawn_command_line_async (APPDIR + '/obmin-center');
            });
            this.requests.connect ('activate', () => {
                GLib.spawn_command_line_async (APPDIR + '/obmin-center');
            });
            this.uploads.connect ('activate', () => {
                GLib.spawn_command_line_async (APPDIR + '/obmin-center');
            });
        }

        this.appmenu.append (new Gtk.SeparatorMenuItem ());
        item = Gtk.MenuItem.new_with_label (_("Control Center..."));
        item.tooltip_text = _("Open Obmin Control Center");
        item.connect ('activate', () => {
            GLib.spawn_command_line_async (APPDIR + '/obmin-center');
        });
        this.appmenu.append (item);

        this.appmenu.append (new Gtk.SeparatorMenuItem ());
        item = Gtk.MenuItem.new_with_label (_("Exit"));
        item.connect ('activate', () => {
            this.remove_events ();
            this.application.quit ();
        });
        this.appmenu.append (item);

        this.appmenu.show_all ();
        this.indicator.set_status (AI.IndicatorStatus.ACTIVE);
        this.indicator.set_icon ("obmin-off");
        this.indicator.set_menu (this.appmenu);
        this.appmenu.connect ('show', () => {
            this.check_status ();
            port = settings.get_int (PORT_KEY);
            this.info_local.update ();
            this.info_public.update ();
            this.update_stats ();
        });
    },

    _enable: function (state) {
        server = state;
        if (state) {
            if (GLib.spawn_command_line_async (APPDIR + "/obmin-server")) {
                this.indicator.set_icon ("obmin-on");
            } else {
                server = false;
                this.indicator.set_icon ("obmin-off");
                this.server_switch.setToggleState (false);
            }
        } else {
            GLib.spawn_command_line_async ("killall obmin-server");
            this.indicator.set_icon ("obmin-off");
        }
        if (this.server_switch.get_active() != server) {
            this.lock = true;
            this.server_switch.set_active (server);
            this.lock = false;
        }
    },

    get server_enabled () {
        let res = GLib.spawn_command_line_sync ("ps -A");
        let o, n;
        if (res[0]) o = res[1].toString().split("\n");
        for (let i = 0; i < o.length; i++) {
            if (o[i].indexOf ("obmin-server") > -1) {
                n = parseInt (o[i].trim().split(" ")[0]);
                if (Number.isInteger(n) && n > 0) return n;
            }
        }
        return false;
    },

    remove_events: function () {
        if (status_event) GLib.Source.remove (status_event);
        status_event = 0;
        if (update_event) GLib.Source.remove (update_event);
        update_event = 0;
        if (this.dbus && this._signalCC) this.dbus.signal_unsubscribe (this._signalCC);
        this._signalCC = 0;
    }
});

const LocalItem = new Lang.Class ({
    Name: 'LocalItem',
    Extends: Gtk.MenuItem,

    _init: function () {
        this.prefix = "<b>" + _("Local IP") + "</b> ";
        this.parent ({label:this.prefix});
        this.get_child().set_markup (this.prefix);
        this.tooltip_text = _("Local Network IP Addresses");
    },

    update_ips: function () {
        let l = Convenience.get_ip_addresses ();
        var scheme = "http://";
        if (https) scheme = "https://";
        this.submenu = new Gtk.Menu ();
        l.forEach ((s)=>{
            let item = Gtk.MenuItem.new_with_label (scheme + s + ":" + port);
            this.submenu.append (item);
        });
        this.get_child().set_markup (this.prefix + l[0]);
        this.submenu.show_all ();
    },

    update: function () {
        this.update_ips ();
    }
});

const PublicItem = new Lang.Class ({
    Name: 'PublicItem',
    Extends: Gtk.MenuItem,

    _init: function () {
        this.prefix = "<b>" + _("Public IP") + "</b> ";
        if (desktop == "PANTHEON") {
            this._ip = Convenience.fetch_sync ("http://ipecho.net/plain", null, null)[0];
            this.parent ({label:_("Public IP") + " " + this._ip});
        } else
            this.parent ({label:this.prefix});
        this.tooltip_text = _("External Network IP Address");
        this._ip = "";
    },

    update: function () {
        Convenience.fetch ("http://ipecho.net/plain", null, null, (text, s) => {
            if ((s == 200) && text) {
                this._ip = text.split("\n")[0];
                if (!this._ip || this._ip.length < 7) this._ip = "";
            } else this._ip = "";
            if (this._ip) this.visible = true;
            else this.visible = false;
            this.get_child().set_markup (this.prefix + this._ip);
            return false;
        });
    }
});

const InfoMenuItem = new Lang.Class ({
    Name: 'InfoMenuItem',
    Extends: Gtk.MenuItem,

    _init: function (desc, val) {
        this.prefix = "<b>" + desc + "</b>  ";
        this.parent ({label:desc});
    },

    update: function (text) {
        this.get_child().set_markup (this.prefix + text);
    }
});

function getCurrentFile () {
    let stack = (new Error()).stack;
    let stackLine = stack.split('\n')[1];
    if (!stackLine)
        throw new Error ('Could not find current file');
    let match = new RegExp ('@(.+):\\d+').exec(stackLine);
    if (!match)
        throw new Error ('Could not find current file');
    let path = match[1];
    let file = Gio.File.new_for_path (path).get_parent();
    return [file.get_path(), file.get_parent().get_path(), file.get_basename()];
}

function get_appdir () {
    let s = getCurrentFile ()[1];
    if (GLib.file_test (s + "/prefs.js", GLib.FileTest.EXISTS)) return s;
    s = GLib.get_home_dir () + "/.local/share/gnome-shell/extensions/obmin@konkor";
    if (GLib.file_test (s + "/prefs.js", GLib.FileTest.EXISTS)) return s;
    s = "/usr/local/share/gnome-shell/extensions/obmin@konkor";
    if (GLib.file_test (s + "/prefs.js", GLib.FileTest.EXISTS)) return s;
    s = "/usr/share/gnome-shell/extensions/obmin@konkor";
    if (GLib.file_test (s + "/prefs.js", GLib.FileTest.EXISTS)) return s;
    throw "Obmin installation not found...";
    return s;
}

let cmd_out, info_out;
function get_info_string (cmd) {
    cmd_out = GLib.spawn_command_line_sync (cmd);
    if (cmd_out[0]) info_out = cmd_out[1].toString().split("\n")[0];
    if (info_out) return info_out;
    return "";
}

function debug (msg) {
    if (msg && (DEBUG > 1)) print ("[obmin][indicator] " + msg);
}

function error (msg) {
    log ("[obmin][indicator] (EE) " + msg);
}

Convenience.initTranslations ();

settings = Convenience.getSettings ();

let app = new ObminIndicator ();
app.application.run (ARGV);
